File: backend/model/problem_data.py
from dataclasses import dataclass
from model.scheme import Scheme
from model.statement import LongStatement


@dataclass
class ProblemData:
    schemeVersions: list[Scheme]
    hypothesis: LongStatement | None

    def serialise(self) -> dict:
        obj = {
            "schemeVersions": [scheme.serialise() for scheme in self.schemeVersions],
        }

        if self.hypothesis is not None:
            obj["hypothesis"] = self.hypothesis.serialise()

        return obj

    @staticmethod
    def deserialise(data: dict) -> "ProblemData":
        scheme = [Scheme.deserialise(scheme) for scheme in data["schemeVersions"]]
        hypothesis = (
            LongStatement.deserialise(data["hypothesis"])
            if "hypothesis" in data
            else None
        )

        return ProblemData(scheme, hypothesis)

File: src/lib/types/problemData.ts
import { serialiseScheme, deserialiseScheme } from "$lib/types/scheme";
import type { Scheme, SchemeSerialised } from "$lib/types/scheme";

export type ProblemData = {
    schemeVersions: Scheme[];
    hypothesis: LongStatement | null;
}

export type ProblemDataSerialsed = {
    schemeVersions: SchemeSerialised[];
    hypothesis: LongStatement | null;
}

export function serialiseProblemData(problemData: ProblemData): ProblemDataSerialsed {
    return {
        schemeVersions: problemData.schemeVersions.map(s => serialiseScheme(s)),
        hypothesis: problemData.hypothesis != null ? problemData.hypothesis : null
    }
}

export function deserialiseProblemData(data: ProblemDataSerialsed): ProblemData {
    return {
        schemeVersions: data.schemeVersions.map(s => deserialiseScheme(s)),
        hypothesis: data.hypothesis != null ? data.hypothesis : null
    }
}

